src/main/java/com/commercetools/sync/producttypes/utils/ProductTypeUpdateAttributeDefinitionActionUtils.java

```diff
@@ -0,0 +1,279 @@
+     * @param newAttributeDefinitionsDrafts                 the new list of attribute definitions drafts.
+     * @param productTypeAttributeDefinitionActionFactory   factory responsible for building attribute definitions
+     *                                                      update actions.
+     * @return a list of attribute definitions update actions if the list of attribute definitions is not identical.
+     *         Otherwise, if the attribute definitions are identical, an empty list is returned.
+     * @throws BuildUpdateActionException in case there are attribute definitions drafts with duplicate names.
+     */
+    @Nonnull
+    private static List<UpdateAction<ProductType>>
+        buildAttributeDefinitionUpdateActionsWithNewAttributeDefinitionDrafts(
+            @Nonnull final List<AttributeDefinition> oldAttributeDefinitions,
+            @Nonnull final List<AttributeDefinitionDraft> newAttributeDefinitionsDrafts,
+            @Nonnull final ProductTypeAttributeActionFactory productTypeAttributeDefinitionActionFactory)
+            throws BuildUpdateActionException {
+
+        final HashSet<String> removedAttributesDefinitionsNames = new HashSet<>();
+
+        final Map<String, AttributeDefinition> oldAttributesDefinitionsNameMap =
+            oldAttributeDefinitions
+                .stream()
+                .collect(toMap(AttributeDefinition::getName, attributeDefinition -> attributeDefinition));
+
+        final Map<String, AttributeDefinitionDraft> newAttributesDefinitionsDraftsNameMap;
+        try {
+            newAttributesDefinitionsDraftsNameMap = newAttributeDefinitionsDrafts.stream().collect(
+                toMap(AttributeDefinitionDraft::getName, attributeDefinitionDraft -> attributeDefinitionDraft,
+                    (attributeDefinitionDraftA, attributeDefinitionDraftB) -> {
+                        throw new DuplicateNameException("Attribute definitions drafts have duplicated names. "
+                            + "Attribute definitions names are expected to be unique inside their product type.");
+                    }
+                ));
+        } catch (final DuplicateNameException exception) {
+            throw new BuildUpdateActionException(exception);
+        }
+
+        final List<UpdateAction<ProductType>> updateActions =
+            buildRemoveAttributeDefinitionOrAttributeDefinitionUpdateActions(
+                oldAttributeDefinitions,
+                removedAttributesDefinitionsNames,
+                newAttributesDefinitionsDraftsNameMap,
+                productTypeAttributeDefinitionActionFactory
+            );
+
+        updateActions.addAll(
+            buildAddAttributeDefinitionUpdateActions(
+                newAttributeDefinitionsDrafts,
+                oldAttributesDefinitionsNameMap,
+                productTypeAttributeDefinitionActionFactory
+            )
+        );
+
+        buildChangeAttributeDefinitionOrderUpdateAction(
```

Comment:
Do you know if it's possible actually through our API that changing the order with names of attributes (even containing names that were just added with **AddAtrributeDefinition** in the same request) would work?

Reply:
I would expect that but in order to be sure, I will try out tomorrow with the playground

Reply:
and? :)

Reply:
oh sorry, I forgot it to respond. yes, it works. I tested by adding a new attribute definition to a product type with just one attribute, and in the same request change the order of both attribute definitions, and it works :)